### DISCUSSION TITLE
Best way to serialize an Interchange to JSON?

### DISCUSSION THREAD

**Discussion by orionarcher** [openforcefield/openff-bespokefit#328] [Q&A] [answered]
Can an Interchange object be serialized without choosing an application-specific representation? It would be convenient for my application to be able to serialize to JSON or a dictionary, like is possible with Molecule and Topology). 

I could use the `to/from OpenMM/Gromacs` as a quick and dirty serialization but that's both application specific and warned against in the docs.

**Comment by pbuslaev** [comment, marked as answer]
Use `Interchange.json()`. This is supported since `Interchange` is inherited from `BaseModel` of [pydantic](https://docs.pydantic.dev/latest/). Though it is not very well [tested](https://github.com/openforcefield/openff-interchange/issues/908)

**Reply by orionarcher** [reply]
Thank you!

**Reply by mattwthompson** [reply]
Pavel is correct - just be slightly more careful than you normally might be, and please let me know if anything looks funky!

**Comment by orionarcher** [comment]
Follow up on this, mostly a question for @mattwthompson. I am developing a MD workflow that will put `Interchange` objects inside the data model of another package (`atomate2`) that uses `pydantic` v2. This is causing a number of issues. I know that `Interchange` 0.4 will switch to `pydantic` v2, but I am wondering:

1. Is there a very rough release date for `Interchange` 0.4? (not looking to hold you to anything, just curious for my development plan).

2. Do you have any suggestions for the meantime? My current solution is to transform the `Interchange` objects to/from dicts whenever they are stored/used.

**Reply by mattwthompson** [reply]
Right now, Interchange uses v2 at installation time even though it uses the v1 API. I thought this  means it can be used alongside other packages that use the v2 API. Are you running into issues (while having v2 installed)?

**Reply by orionarcher** [reply]
I am using v2 and I didn't run into problems until I tried to use `Interchange` as a member of a class inheriting from `pydantic v2 `. For example:

```python
from openff.interchange import Interchange
from pydantic import BaseModel # importing v2 basemodel

class A(BaseModel):
    interchange: Interchange

interchange = ... # somehow define interchange

A(interchange=interchange)
```

The error I am encountering is documented here: https://github.com/pydantic/pydantic/issues/7997. It seems to be a feature of v1 vs v2, not a bug.

**Reply by mattwthompson** [reply]
Ah, okay ... I hadn't tried that. That's really unfortunate. I don't have a better solution off the top of my head besides using an older version of your other upstreams or vendor some frankenstein combination of things together.

Just to clarify, 0.4 will be v2-only at installation-time, it'll still be using the v1 API. The migration is blocked by other parts of our infrastructure, and unfortunately I don't have a timeline for completing it. Hopefully this summer @Yoshanuikabundi and/or myself will have the time to tear through it!
* https://github.com/openforcefield/openff-models/pull/26
* https://github.com/openforcefield/openff-units/pull/77

**Reply by orionarcher** [reply]
Got it, thank you! Storing dicts is a workable solution in the meantime. Looking forward to `0.4`.
